consensus: Remove special treatment for single threaded script checking

fjahr opened the pull request:
This topic has been motivated by my work on batch validation and a related conversation just happened here: https://github.com/bitcoin/bitcoin/pull/32467#discussion_r2098305199 

`CheckInputScripts` currently only does what its name implies if there is no multithreading usage for validation. If there are worker threads available the function instead only creates the checks and puts them into a vector. Aside from some shared pre-checks the multithreaded code path is much simpler compared to the rest. This dual use makes the function hard to grasp and its naming confusing.

This PR refactors this code to be more readable and improves documentation aside from unifying the single threaded and the multithreaded cases by letting checkqueue handle single threaded script checking as well. The code is already suited for this and there seems to be not noticable performance overhead from this.

fjahr committed:
refactor, validation: Extract EnsureTxData from CheckInputScripts

Move the spent outputs initialization and txdata setup into a
standalone helper.

theuni commented:
Concept ACK. Anything to clean this up :)

See also #32317 which moves some of this functionality around similarly.

fjahr commented:
Cleaned up the comments so this should be ready for more detailed feedback, but expect a rebase once https://github.com/bitcoin/bitcoin/pull/32467 is merged.

fjahr commented on src/validation.cpp:
```diff
@@ -2185,16 +2189,46 @@ void EnsureTxData(const CTransaction& tx, const CCoinsViewCache& inputs, Precomp
+                       bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
+                       ValidationCache& validation_cache) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    if (tx.IsCoinBase()) return {};
```
I have been contemplating kicking this out (here and in `CheckInputScripts`) or even turning this into an assert because the callers already handle this everywhere. On the other hand it conceptually makes sense to return early if we know there is nothing to check because we don't expect any inputs. So I am still undecided on this one.

TheCharlatan commented:
Concept ACK

stickies-v commented:
Concept ACK

theuni commented:
Ready for rebase.

fjahr commented:
> Ready for rebase.

Done, ready for review :)

theuni commented on src/validation.cpp:
```diff
@@ -2695,22 +2722,20 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
-                state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
-                              tx_state.GetRejectReason(), tx_state.GetDebugMessage());
-                break;
+                if (!CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache)) {
```
I think this should go a little further. At this point, `CheckInputScripts` is essentially meant to be a mempool function and `PrepareInputScriptChecks` is for validation, except in the weird edge-case  when we're single-threaded.

I think I'd rather just see this go all the way and make the split purposeful. Something like:
```patch
diff --git a/src/validation.cpp b/src/validation.cpp
index 1c81620889f..d9a447b460f 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2723,19 +2723,18 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
             TxValidationState tx_state;
-            // If parallel script checking is possible (worker threads are available) the checks are appended to a
-            // vector without running them. The vector is then added to control which runs the checks asynchronously.
-            // Otherwise, CheckInputScripts runs the checks on a single thread before returning.
-            if (control) {
-                std::vector<CScriptCheck> vChecks = PrepareInputScriptChecks(tx, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
-                control->Add(std::move(vChecks));
-            } else {
-                if (!CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache)) {
-                    // Any transaction validation failure in ConnectBlock is a block consensus failure
-                    state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
-                                  tx_state.GetRejectReason(), tx_state.GetDebugMessage());
-                    break;
+            // If parallel script checking is possible (worker threads are available), they are added to control which
+            // which runs the checks asynchronously. Otherwise they are run here directly.
+            std::vector<CScriptCheck> vChecks = PrepareInputScriptChecks(tx, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
+            if (control) control->Add(std::move(vChecks));
+            else {
+                for (auto& check : vChecks) {
+                    if (const auto& check_result = check()) {
+                        state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, strprintf("mandatory-script-verify-flag-failed (%s)", ScriptErrorString(check_result->first)), check_result->second);
+                        break;
+                    }
                 }
+                if (!state.IsValid()) break;
             }
         }

```

An alternative would be to make `CCheckQueueControl` work with 0 worker threads (maybe it already does?) and just always use it. That would completely unify the approach here.

mzumsande replied:
I like that idea.

I think it would be a small change in behaviour, because  if I read [this line](https://github.com/bitcoin/bitcoin/blob/2df824f4e62b6bc569044819cd64f66f3839ba13/src/validation.cpp#L2248) right it will add entries to the cache if 
1)`pvChecks` is not set (single-threaded) 
2) `cacheFullScriptStore` is set (i.e. if `ConnectBlock()` was called with `fJustCheck`, so currently only from `TestBlockValidity`).
I wonder if the current behavior is intended - it seems very strange to me that mining rpcs would add entries to the script cache, but only in the `-par=1` case.

theuni replied:
@mzumsande I think it's very likely that that behavior is intended :)

fjahr replied:
Took the suggestion here.

mzumsande commented on src/validation.cpp:
```diff
@@ -2185,16 +2189,46 @@ void EnsureTxData(const CTransaction& tx, const CCoinsViewCache& inputs, Precomp
+
+    EnsureTxData(tx, inputs, txdata);
+
+    std::vector<CScriptCheck> pvChecks{};
```
Just a rough idea, but did you consider the alternative approach of doing the `pvChecks` population not here (but in `ConnectBlock()` or a separate function), so that `CheckInputScripts` could also call `PrepareInputScriptChecks` and there would be no duplication of the coinbase check, cache check and `TxData` init?

sedited commented on src/validation.cpp:
```diff
@@ -2680,22 +2672,30 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
-                              tx_state.GetRejectReason(), tx_state.GetDebugMessage());
-                break;
+
+            if (control) control->Add(std::move(vChecks));
```
Can you add braces here? I find different inlining styles for the same if-else block slightly confusing to read.

sedited commented:
Looking at this patch, I am asking myself why we're keeping the single threaded / no checkqueue case in the first place. This seems to introduce a vector of jobs for both cases now, so maybe we could just create a queue without any workers?

sedited reviewed (commented):
Approach ACK

This seems like a nice cleanup, will look at it in detail later.

l0rinc commented on src/validation.cpp:
```diff
@@ -2068,6 +2068,22 @@ bool IsScriptCached(ValidationCache& validation_cache, bool cacheFullScriptStore
     return false;
 }
 
+void EnsureTxData(const CTransaction& tx, const CCoinsViewCache& inputs, PrecomputedTransactionData& txdata)
```
> [6b37953](https://github.com/bitcoin/bitcoin/pull/32575/changes/6b379531347eb6715ad8373a2311a6811b36f17e) _refactor: Extract TxData initialization from CheckInputScripts_:

This new abstraction is doing multiple things (which makes it a bit hard to name):
* asking `txdata` whether the payload will be accepted to avoid calculating `GetSpentOutputs` needlessly
* calculating it if it will be accepted
* passing it to `txdata.Init`
* asserting that everything makes sense regardless of previous actions

Most of these make more sense at the call site. The only excessive calculation is the to-be-spent outputs calculation. We could extract only that (making sure no extra copies happen, of course), something like:
```patch
diff --git a/src/validation.cpp b/src/validation.cpp
--- a/src/validation.cpp	(revision d26ff1ff03da3531d1a6a28466784911ece28e0a)
+++ b/src/validation.cpp	(date 1772711510934)
@@ -2048,6 +2048,18 @@
     m_script_execution_cache.insert(entry);
 }
 
+static std::vector<CTxOut> GetSpentOutputs(const CTransaction& tx, const CCoinsViewCache& inputs)
+{
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(tx.vin.size());
+    for (const auto& txin : tx.vin) {
+        const Coin& coin = inputs.AccessCoin(txin.prevout);
+        assert(!coin.IsSpent());
+        spent_outputs.emplace_back(coin.out);
+    }
+    return spent_outputs;
+}
+
 /**
  * Check whether all of this transaction's input scripts succeed.
  *
@@ -2091,18 +2103,7 @@
         return true;
     }
 
-    if (!txdata.m_spent_outputs_ready) {
-        std::vector<CTxOut> spent_outputs;
-        spent_outputs.reserve(tx.vin.size());
-
-        for (const auto& txin : tx.vin) {
-            const COutPoint& prevout = txin.prevout;
-            const Coin& coin = inputs.AccessCoin(prevout);
-            assert(!coin.IsSpent());
-            spent_outputs.emplace_back(coin.out);
-        }
-        txdata.Init(tx, std::move(spent_outputs));
-    }
+    if (!txdata.m_spent_outputs_ready) txdata.Init(tx, GetSpentOutputs(tx, inputs));
     assert(txdata.m_spent_outputs.size() == tx.vin.size());
 
     for (unsigned int i = 0; i < tx.vin.size(); i++) {
```

https://godbolt.org/z/jj3eGq8eY indicates that `GetSpentOutputs` allocates once via `reserve`, and `Init`'s `&&` parameter steals the buffer pointer directly - so it's a pure refactor.

l0rinc reviewed (changes requested):
Concept ACK. The recent consensus bug in the BIP110 soft fork proposal in this exact area indicates the current code is hard to understand.                                                    
Approach NACK. The recent consensus bug in the BIP110 soft fork proposal in this exact area indicates we need to be extra careful, which isn't the case in this PR yet.                       

I left a ton of suggestions as I recreated the change locally, experimenting with a few different implementations. We should definitely split the two intertwined functionalities, but not in the same PR that also modifies the behavior. There are too many unexplained changes here that I strongly object to.                                                                        

Let's untangle the two functionalities first. That refactor is verbose but quite safe. If it gains momentum and gets merged, we can build on it to unify the two even more, if needed.

Edit:
I measured multithreaded script verification, and I can confirm that it doesn't introduce a slowdown
<details><summary>Benchmark results</summary>

```python
for DBCACHE in 5000; do \
    COMMITS="043301a537fdad469ef83ee3d1033404a0cefe16 7726a6a6c3e378e538ecb4b5bfd82c2880cc0688"; \
    STOP=936639; CC=gcc; CXX=g++; \
    BASE_DIR="/mnt/my_storage"; DATA_DIR="$BASE_DIR/BitcoinData"; LOG_DIR="$BASE_DIR/logs"; \
    (echo ""; for c in $COMMITS; do git fetch -q origin $c && git log -1 --pretty='%h %s' $c || exit 1; done) && \
    (echo "" && echo "$(date -I) | reindex-chainstate | ${STOP} blocks | dbcache ${DBCACHE} | $(hostname) | $(uname -m) | $(lscpu | grep 'Model name' | head -1 | cut -d: -f2 | xargs) | $(nproc) cores | $(free -h | awk '/^Mem:/{print $2}') RAM | $(lsblk -no ROTA $(df --output=source $BASE_DIR | tail -1) | grep -q 1 && echo HDD || echo SSD)"; echo "") && \
    hyperfine \
    --sort command \
    --runs 1 \
    --export-json "$BASE_DIR/rdx-$(sed -E 's/(\w{8})\w+ ?/\1-/g;s/-$//'<<<"$COMMITS")-$STOP-$DBCACHE-$CC.json" \
    --parameter-list COMMIT ${COMMITS// /,} \
    --prepare "killall -9 bitcoind 2>/dev/null; rm -f ./build/bin/bitcoind; git clean -fxd; git reset --hard {COMMIT} && \
      cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DENABLE_IPC=OFF && ninja -C build bitcoind -j1 && \
      ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=1000 -printtoconsole=0; sleep 20; rm -f $DATA_DIR/debug.log" \
    --conclude "killall bitcoind || true; sleep 5; grep -q 'height=0' $DATA_DIR/debug.log && grep -q 'Disabling script verification at block #1' $DATA_DIR/debug.log && grep -q 'height=$STOP' $DATA_DIR/debug.log && grep 'Bitcoin Core version' $DATA_DIR/debug.log | grep -q "$(printf %.12s {COMMIT})"; \
                cp $DATA_DIR/debug.log $LOG_DIR/debug-{COMMIT}-$(date +%s).log" \
    "COMPILER=$CC ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=$DBCACHE -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0";
done

043301a537 refactor: Extract cache check from CheckInputScripts
7726a6a6c3 validation, refactor: Remove single threaded special case

2026-03-05 | reindex-chainstate | 936639 blocks | dbcache 5000 | i9-ssd | x86_64 | Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz | 16 cores | 62Gi RAM | SSD

Benchmark 1: COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 (COMMIT = 043301a537fdad469ef83ee3d1033404a0cefe16)
  Time (abs ≡):        37023.077 s               [User: 389375.279 s, System: 1125.389 s]
 
Benchmark 2: COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 (COMMIT = 7726a6a6c3e378e538ecb4b5bfd82c2880cc0688)
  Time (abs ≡):        37195.482 s               [User: 388819.276 s, System: 1134.977 s]
 
Relative speed comparison
        1.00          COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 (COMMIT = 043301a537fdad469ef83ee3d1033404a0cefe16)
        1.00          COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 (COMMIT = 7726a6a6c3e378e538ecb4b5bfd82c2880cc0688)
```

</details>

I will measure single-threaded version next.

fjahr committed:
refactor, validation: Encapsulate script execution cache access in ValidationCache

Move m_script_execution_cache to private and expose IsScriptValidated()
and CacheScriptValidation() methods. This ensures the AssertLockHeld
check is consistently applied for both cache lookups and insertions,
rather than relying on callers to assert the lock.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

fjahr committed:
refactor: Extract GetScriptCacheEntry from CheckInputScripts

Separate the script execution cache lookup (coinbase check and cache
lookup) into GetScriptCacheEntry. It returns std::optional<uint256> —
nullopt if no script checks are needed (coinbase or cache hit), or the
cache entry hash for use by the caller when storing results later.

CheckInputScripts now calls GetScriptCacheEntry internally and
initializes txdata via EnsureTxData only when checks are needed. This
prepares for ConnectBlock to call GetScriptCacheEntry directly and
construct script checks without going through CheckInputScripts.

fjahr committed:
test: Use GetScriptCacheEntry in txvalidationcache_tests

Update cache-hit and cache-miss tests to call GetScriptCacheEntry
directly instead of checking the pvChecks vector populated by
CheckInputScripts. Both functions use the same caching logic
internally, and GetScriptCacheEntry is the more direct way to test
whether a transaction's scripts would need validation.

fjahr committed:
validation: Use checkqueue for all script checking in ConnectBlock

Instead of having two code paths in ConnectBlock, one that collects
script checks for the checkqueue (multithreaded) and one that runs
CheckInputScripts inline (single-threaded), just always use the
checkqueue. ConnectBlock now calls GetScriptCacheEntry directly and
constructs CScriptCheck objects to add to the queue regardless of the
thread count.

This removes the HasThreads() check and the single-threaded special
case. The checkqueue already handles the case of having zero worker
threads by running checks on the calling thread.

Previously, script execution cache population during TestBlockValidity
(fJustCheck=true) only worked in the single-threaded case, because
CheckInputScripts guarded the cache insertion with !pvChecks. In the
multithreaded case, pvChecks was always set, so the cache was never
populated. This inconsistency has existed since the script execution
cache was introduced. Now that ConnectBlock no longer uses
CheckInputScripts, cache population is done explicitly after it is
confirmed that all script checks have passed. This means we populate
the cache in both single-threaded and multithreaded cases.

fjahr committed:
refactor: Remove pvChecks parameter from CheckInputScripts

Since ConnectBlock now constructs script checks directly via
GetScriptCacheEntry, no caller passes a pvChecks vector to
CheckInputScripts. Remove the parameter and the associated
dead code paths.

l0rinc commented:
> <img width="593" height="115" alt="image" src="https://github.com/user-attachments/assets/9507a5b9-0445-4e9c-b465-d919ac31ea21" />

A week later and this is still running against your "refactor"

l0rinc commented:
> No measurable reduction of speed after the change. Benchmark-wise, the PR is good

Thanks for checking. I ran it again and it crashed my server again. It's possible it's unrelated to this PR, will retry on a different machine, maybe the server is toast. Or maybe there's a race condition.
Doesn't really matter, my main concerns are still:
* The PR title still says "refactor" but commit d84cbf4b5366d4802f973f99e65c903daad80923 unifies the single-threaded and multithreaded paths, which is a behavior change in consensus-critical code. I still have [a hard time trusting this completely](https://github.com/bitcoin/bitcoin/pull/32575#discussion_r2891176328), I think we should do the refactor and the behavior change [separately](https://github.com/bitcoin/bitcoin/pull/34875).
* The back-and-forth around cache population (e.g. darosior's [correction](https://github.com/bitcoin/bitcoin/pull/32575#issuecomment-4075771363)) illustrates exactly why structural refactoring and behavioral changes shouldn't be done in the same PR for consensus critical code.
* Missing upfront test coverage to demonstrate no behavior change - see my alternative which [adds thread-mode tests first](https://github.com/bitcoin/bitcoin/pull/34875/commits/26715c2b887e) before any refactoring, so regressions are caught immediately. I'd be happy to close that PR if you decide to do only the refactoring here, but I don't like the behavior changes here.

Edit:
I managed to run `-par=1` on a different server, it has very similar performance. My objections still stand, see above.

<details><summary>Before</summary>

```bash
for DBCACHE in 5000; do \
    COMMITS="544c15ff4e22e0271f91341b446ad84b6f4fd6c3"; \
    STOP=936639; CC=gcc; CXX=g++; \
    BASE_DIR="/mnt/my_storage"; DATA_DIR="$BASE_DIR/BitcoinData"; LOG_DIR="$BASE_DIR/logs"; \
    (echo ""; for c in $COMMITS; do git fetch -q origin $c && git log -1 --pretty='%h %s' $c || exit 1; done) && \
    (echo "" && echo "$(date -I) | reindex-chainstate | ${STOP} blocks | dbcache ${DBCACHE} | $(hostname) | $(uname -m) | $(lscpu | grep 'Model name' | head -1 | cut -d: -f2 | xargs) | $(nproc) cores | $(free -h | awk '/^Mem:/{print $2}') RAM | $(lsblk -no ROTA $(df --output=source $BASE_DIR | tail -1) | grep -q 1 && echo HDD || echo SSD)"; echo "") && \
    hyperfine \
    --sort command \
    --runs 1 \
    --export-json "$BASE_DIR/rdx-$(sed -E 's/(\w{8})\w+ ?/\1-/g;s/-$//'<<<"$COMMITS")-$STOP-$DBCACHE-$CC.json" \
    --parameter-list COMMIT ${COMMITS// /,} \
    --prepare "killall -9 bitcoind 2>/dev/null; rm -f ./build/bin/bitcoind; git clean -fxd; git reset --hard {COMMIT} && \
      cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DENABLE_IPC=OFF && ninja -C build bitcoind -j1 && \
      ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=1000 -printtoconsole=0; sleep 20; rm -f $DATA_DIR/debug.log" \
    --conclude "killall bitcoind || true; sleep 5; grep -q 'height=0' $DATA_DIR/debug.log && grep -q 'Disabling script verification at block #1' $DATA_DIR/debug.log && grep -q 'height=$STOP' $DATA_DIR/debug.log && grep 'Bitcoin Core version' $DATA_DIR/debug.log | grep -q "$(printf %.12s {COMMIT})"; \
                cp $DATA_DIR/debug.log $LOG_DIR/debug-{COMMIT}-$(date +%s).log" \
    "COMPILER=$CC ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=$DBCACHE -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 -par=1";
done

544c15ff4e Merge bitcoin/bitcoin#34759: walletdb: hash pubkey/privkey in one shot to avoid leaking secret data

2026-04-10 | reindex-chainstate | 936639 blocks | dbcache 5000 | i7-hdd | x86_64 | Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz | 8 cores | 62Gi RAM | HDD

Benchmark 1: COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 -par=1 (COMMIT = 544c15ff4e22e0271f91341b446ad84b6f4fd6c3)
  Time (abs ≡):        223862.666 s               [User: 220249.873 s, System: 1347.124 s]
```

</details>


<details><summary>After</summary>

```bash
for DBCACHE in 5000; do \
    COMMITS="23ac5f3f48d00a2087f9676609eb2e4b48813235"; \
    STOP=936639; CC=gcc; CXX=g++; \
    BASE_DIR="/mnt/my_storage"; DATA_DIR="$BASE_DIR/BitcoinData"; LOG_DIR="$BASE_DIR/logs"; \
    (echo ""; for c in $COMMITS; do git fetch -q origin $c && git log -1 --pretty='%h %s' $c || exit 1; done) && \
    (echo "" && echo "$(date -I) | reindex-chainstate | ${STOP} blocks | dbcache ${DBCACHE} | $(hostname) | $(uname -m) | $(lscpu | grep 'Model name' | head -1 | cut -d: -f2 | xargs) | $(nproc) cores | $(free -h | awk '/^Mem:/{print $2}') RAM | $(lsblk -no ROTA $(df --output=source $BASE_DIR | tail -1) | grep -q 1 && echo HDD || echo SSD)"; echo "") && \
    hyperfine \
    --sort command \
    --runs 1 \
    --export-json "$BASE_DIR/rdx-$(sed -E 's/(\w{8})\w+ ?/\1-/g;s/-$//'<<<"$COMMITS")-$STOP-$DBCACHE-$CC.json" \
    --parameter-list COMMIT ${COMMITS// /,} \
    --prepare "killall -9 bitcoind 2>/dev/null; rm -f ./build/bin/bitcoind; git clean -fxd; git reset --hard {COMMIT} && \
      cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DENABLE_IPC=OFF && ninja -C build bitcoind -j1 && \
      ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=1000 -printtoconsole=0; sleep 20; rm -f $DATA_DIR/debug.log" \
    --conclude "killall bitcoind || true; sleep 5; grep -q 'height=0' $DATA_DIR/debug.log && grep -q 'Disabling script verification at block #1' $DATA_DIR/debug.log && grep -q 'height=$STOP' $DATA_DIR/debug.log && grep 'Bitcoin Core version' $DATA_DIR/debug.log | grep -q "$(printf %.12s {COMMIT})"; \
                cp $DATA_DIR/debug.log $LOG_DIR/debug-{COMMIT}-$(date +%s).log" \
    "COMPILER=$CC ./build/bin/bitcoind -datadir=$DATA_DIR -stopatheight=$STOP -dbcache=$DBCACHE -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 -par=1";
done

23ac5f3f48 refactor: Remove pvChecks parameter from CheckInputScripts

2026-04-07 | reindex-chainstate | 936639 blocks | dbcache 5000 | i7-hdd | x86_64 | Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz | 8 cores | 62Gi RAM | HDD

Benchmark 1: COMPILER=gcc ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=936639 -dbcache=5000 -reindex-chainstate -blocksonly -connect=0 -printtoconsole=0 -assumevalid=0 -par=1 (COMMIT = 23ac5f3f48d00a2087f9676609eb2e4b48813235)
  Time (abs ≡):        230301.470 s               [User: 219951.017 s, System: 1338.768 s]
```

</details>

fjahr commented:
> I managed to run -par=1 on a different server, it has very similar performance.

Thanks for finally confirming that.

For the future: Maybe post a new comment with your benchmark results instead of editing a comment that is over a week old. This way, other reviewers will get a notification which may trigger them to continue their review. Review which might have been blocked by your earlier comments about a potential performance issue.

> The PR title still says "refactor" but commit https://github.com/bitcoin/bitcoin/commit/d84cbf4b5366d4802f973f99e65c903daad80923 unifies the single-threaded and multithreaded paths, which is a behavior change in consensus-critical code. I still have https://github.com/bitcoin/bitcoin/pull/32575#discussion_r2891176328, I think we should do the refactor and the behavior change https://github.com/bitcoin/bitcoin/pull/34875.

I have removed `refactoring` from the title although I think it's highly debatable if this is necessary. It seems like by your definition no code change is actually a refactor unless, maybe, it compiles to the exact same ASM. The way consensus rules are enforced is still exactly the same, there isn't even new code being used for it, it's using the same code just without worker threads which already works because the master thread has always been participating in the work. Thus this is still a refactor by my definition.

> The back-and-forth around cache population (e.g. darosior's https://github.com/bitcoin/bitcoin/pull/32575#issuecomment-4075771363) illustrates exactly why structural refactoring and behavioral changes shouldn't be done in the same PR for consensus critical code.

Indeed @darosior made a great observation and I made an according change but as you can see from the prior comments, the code was previously broken and the benefit of this is not entirely clear either so this would not have been a dramatic issue either way it seems. I can also promise you that I would have likely made the same mistake if I had broken up the PR in two from the start and we don't know if @darosior would have caught the issue in that alternative universe. So, all the evidence that I see is that the review process works on this PR.

> Missing upfront test coverage to demonstrate no behavior change - see my alternative which [adds thread-mode tests first](https://github.com/bitcoin/bitcoin/pull/34875/commits/26715c2b887e) before any refactoring, so regressions are caught immediately. I'd be happy to close that PR if you decide to do only the refactoring here, but I don't like the behavior changes here.

The functional tests [already have `-par=1` coverage](https://github.com/bitcoin/bitcoin/blob/master/test/functional/p2p_segwit.py#L222) and the fuzz test also use 0 worker threads in deterministic mode so I think this doesn't make a huge difference but if this makes you happy finally, I can pull the tests commit in here since it doesn't seem to hurt at first glance. Otherwise you could also add it as a follow-up or open a stand-alone PR. Since your PR changes a similar number of LoCs I don't think it's any easier to review than this PR here and just because you don't use CheckQueue it isn't any less "dangerous". So stacking another PR on top of yours then to get what we actually want isn't helpful in terms of safety or review burden, but I have already laid this out previously in more detail. It also doesn't seem like your comments have changed the minds of the other, previous commenters, so I will leave the changes here as is.

l0rinc commented:
> Maybe post a new comment with your benchmark results instead of editing a comment that is over a week old

I have explicitly pinged you and the original comment already stated that performance wasn't my main concern and that I was still measuring it on a different machine. Are you not curious why the other server failed?
The updated results don't change my review - my objections are about mixing a behavior change with a refactor in consensus-critical code, not performance.

> The way consensus rules are enforced is still exactly the same

I disagree. The [BIP-110 consensus failure](https://github.com/bitcoinknots/bitcoin/pull/238#issuecomment-3896624715) specifically relied on the `-par=1` path, which is exactly the path being unified here.
That bug demonstrates concretely that changes in this area can have non-obvious consensus implications - which is why I keep suggesting we separate the behavior change from the pure refactor.

> just because you don't use CheckQueue it isn't any less "dangerous"

I think it is, precisely because of the above. Keeping both paths intact while cleaning up their implementations avoids introducing new interactions between previously independent code paths.

> but if this makes you happy finally

I'm raising these concerns because I think they matter for the safety of the codebase, and I find the overall dismissive tone of this response concerning. We don't have to agree, but I'd appreciate the concerns being engaged with on their technical merits.

> Since your PR changes a similar number of LoCs

The line count isn't the concern. It's combining a behavior change with a refactor in consensus-critical code. darosior's [correction](https://github.com/bitcoin/bitcoin/pull/32575#issuecomment-4075771363) about cache population and the BIP-110 consensus bug are concrete examples of why that combination is risky - that issue was caught only after detailed review, and it directly resulted from mixing refactoring with behavioral changes. Separating the two would make each part individually easier to verify.

fjahr commented:
> Are you not curious why the other server failed?

I am more than happy to look it when you actually post some information on what might have happened there. Until then it's just a distraction and I don't think this PR should be kept hostage by it after there was no follow-up for a long time. The performance issue that you were threatening for weeks didn't turn out to be a problem either.

> The updated results don't change my review

That's fine, I think I have addressed your comments suffiently, taking into account earlier reviews and weighing them against your concerns as well as my personal view and laying out my arguments as clearly as I can. But I am always happy to hear from other reviewers and reconsider if they have changed their mind based on your comments. I will not engage further on you posting your earlier arguments repeatedly since it seems like a waste of everyone's time.

> We don't have to agree

Right, let's move on.

fjahr commented:
Rebase after #35209

Eunovo commented on src/validation.cpp:
```diff
@@ -2065,6 +2070,32 @@ static void EnsureTxData(const CTransaction& tx, const CCoinsViewCache& inputs,
+ *
+ * Non-static (and redeclared) in src/test/txvalidationcache_tests.cpp
+ */
+std::optional<uint256> PrepareScriptChecks(const CTransaction& tx,
```
https://github.com/bitcoin/bitcoin/pull/32575/commits/74a05c8b3b318aa4406bf0cfca9bf155bb1bbb71:

I think a better name for `PrepareScriptChecks` is `GetScriptCacheKey` or `GetScriptCacheEntry`, since it returns the Script Cache key if the script is not already validated. I find `PrepareScriptChecks` to be a confusing name for what it actually does.

fjahr replied:
A pure renaming would have dropped the "prepare" part (`EnsureTxData`) under the table, which IMO would not have been better than the status quo. But since @l0rinc also had a similar issue with this function that I didn't address back then, I now moved `EnsureTxData` out of this function and into the call sites. This means I can apply your renaming suggestion cleanly to the remaining function here.

Eunovo commented on src/validation.cpp:
```diff
@@ -2648,6 +2645,11 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         LogInfo("Block validation error: %s", state.ToString());
         return false;
     }
+    // Cache script execution results when doing TestBlockValidity (fJustCheck=true),
```
https://github.com/bitcoin/bitcoin/pull/32575/commits/1bfa2ec4bb3bd27a8fd4bf276b6931a010797294:

I found this comment to be confusing. The actual decision point where `script_cache_entries` is populated is several lines above; `fJustCheck=true` is not relevant to this loop. I think it is much better to remove this comment.

A better comment here should state that we only cache entries after we determine that validation succeeded.

fjahr replied:
Yeah, makes sense, I moved this comment with minor edits and added a new comment here per your suggestion.

Eunovo reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/32575/commits/dfecae46c14c98e62ad8e2503493b152cb9b5b91:

The changes simplify the flow for script verification in `ConnectBlock`, and they have a nice side effect of simplifying `txvalidationcache_tests.cpp`.

fjahr commented:
Addressed @Eunovo 's comments, thanks for the review!
